app/assets/javascripts/oxalis/model/binary/texture_bucket_manager.js

```diff
@@ -0,0 +1,283 @@
+    const lookUpBufferSize = Math.pow(lookUpBufferWidth, 2) * floatsPerLookUpEntry;
+    this.bucketsPerDim = bucketsPerDim;
+    this.textureWidth = textureWidth;
+    this.dataTextureCount = dataTextureCount;
+
+    this.lookUpBuffer = new Float32Array(lookUpBufferSize);
+    this.freeIndexSet = new Set(_.range(this.bufferCapacity));
+
+    this.dataTextures = [];
+
+    this.keepLookUpBufferUpToDate();
+    this.processWriterQueue();
+  }
+
+  clear() {
+    this.setActiveBuckets([], [0, 0, 0, 0], [0, 0, 0, 0]);
+  }
+
+  freeBucket(bucket: DataBucket): void {
+    const unusedIndex = this.activeBucketToIndexMap.get(bucket);
+    if (unusedIndex == null) {
+      return;
+    }
+    this.activeBucketToIndexMap.delete(bucket);
+    this.committedBucketSet.delete(bucket);
+    this.freeIndexSet.add(unusedIndex);
+  }
+
+  // Takes an array of buckets (relative to an anchorPoint) and ensures that these
+  // are written to the dataTexture. The lookUpTexture will be updated to reflect the
+  // new buckets.
+  setActiveBuckets(
+    buckets: Array<DataBucket>,
+    anchorPoint: Vector4,
+    fallbackAnchorPoint: Vector4,
+  ): void {
+    this.currentAnchorPoint = anchorPoint;
+    this.fallbackAnchorPoint = fallbackAnchorPoint;
+    // Find out which buckets are not needed anymore
+    const freeBucketSet = new Set(this.activeBucketToIndexMap.keys());
+    for (const bucket of buckets) {
+      freeBucketSet.delete(bucket);
+    }
+
+    // Remove unused buckets
+    const freeBuckets = Array.from(freeBucketSet.values());
+    for (const freeBucket of freeBuckets) {
+      this.freeBucket(freeBucket);
+    }
+
+    const freeIndexArray = Array.from(this.freeIndexSet);
+    for (const nextBucket of buckets) {
+      if (!this.activeBucketToIndexMap.has(nextBucket)) {
+        if (freeIndexArray.length === 0) {
+          throw new Error("A new bucket should be stored but there is no space for it?");
+        }
+        const freeBucketIdx = freeIndexArray.shift();
+        this.reserveIndexForBucket(nextBucket, freeBucketIdx);
+      }
+    }
+
+    this._refreshLookUpBuffer();
+  }
+
+  keepLookUpBufferUpToDate() {
+    if (this.isRefreshBufferOutOfDate) {
+      this._refreshLookUpBuffer();
+    }
+    window.requestAnimationFrame(() => {
+      this.keepLookUpBufferUpToDate();
+    });
+  }
+
+  // Commit "active" buckets by writing these to the dataTexture.
+  processWriterQueue() {
+    let processedItems = 0;
+    // uniqBy removes multiple write-buckets-requests for the same index.
+    // It preserves the first occurence of each duplicate, which is why
+    // this queue has to be filled from the front (via unshift) und read from the
+    // back (via pop). This ensures that the newest bucket "wins" if there are
+    // multiple buckets for the same index.
+    this.writerQueue = _.uniqBy(this.writerQueue, el => el._index);
+    const maxBucketCommitsPerFrame = 30;
```

Comment:
Yes, I determined it empirically. However, this obviously depends on the machine. I'll rewrite it so that the loop stays busy until 16ms have passed. We could play with the value of 16ms a bit, but I think that's a reasonable default.

Reply:
I played a bit around with the value and committed the version with 16ms but I couldn't really see a big difference. Maybe you can also fiddle with it to see what works best on your hardware.